### PR TITLE
[Data] Allow fusing MapOperator -> AllToAllOperator

### LINKED ISSUE #32608: [Datasets] Allow fuse MapOperator with AllToAllOperator
### Description

Currently we only support operator fusion for MapOperator. We should also support fusing MapOperator with AllToAllOperator (e.g. map_batches and shuffle map).

### Use case

_No response_

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?
Currently, we only support operator fusion for MapOperator. We should also support fusing MapOperator with AllToAllOperator; this PR allows for one direction, MapOperator -> AllToAllOperator. In a future PR, we can work on the other direction, AllToAllOperator -> MapOperator.

## Related issue number

Closes #32608 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py [hunk 1]
```diff
@@ -53,12 +53,13 @@ class RandomShuffle(AbstractAllToAll):
...
         self,
         input_op: LogicalOperator,
+        name: str = "RandomShuffle",
         seed: Optional[int] = None,
         num_outputs: Optional[int] = None,
         ray_remote_args: Optional[Dict[str, Any]] = None,
     ):
         super().__init__(
-            "RandomShuffle",
+            name,
```
Comment (scottjlee) [inline python/ray/data/_internal/logical/operators/all_to_all_operator.py:62]: this is needed so we can override the name with the upstream operator-fused name.

File: python/ray/data/_internal/logical/rules/operator_fusion.py [hunk 2]
```diff
@@ -205,6 +262,56 @@ def transform_fn(blocks: Iterator[Block], ctx: TaskContext) -> Iterator[Block]:
...
+
+        op = AllToAllOperator(
+            fused_transform_fn_map_alltoall,
+            input_op,
+            name=name,
+        )
+        # Bottom out at the source logical op (e.g. Read()).
+        input_op = up_logical_op
+
+        logical_op = RandomShuffle(input_op, name=name, ray_remote_args=ray_remote_args)
```
Comment (c21) [inline python/ray/data/_internal/logical/rules/operator_fusion.py:310]: how do we support other operator like `Repartition`?

File: python/ray/data/_internal/execution/interfaces.py [hunk 3]
```diff
@@ -233,6 +233,10 @@ class TaskContext:
     # TODO(chengsu): clean it up from TaskContext with new optimizer framework.
     sub_progress_bar_dict: Optional[Dict[str, ProgressBar]] = None
 
+    # The underlying function called in a MapOperator; this is used when fusing
+    # an AllToAllOperator with an upstream MapOperator.
+    map_transform_fn: Optional["MapTransformFn"] = None
```
Comment (c21) [inline python/ray/data/_internal/execution/interfaces.py:238]: can we call it `upstream_map_transform_fn` or `input_map_transform_fn`? We will soon add support to fuse downstream map operators later.

File: python/ray/data/tests/test_execution_optimizer.py [hunk 4]
```diff
@@ -607,6 +607,33 @@ def __call__(self, x):
     assert isinstance(physical_op.input_dependencies[0], InputDataBuffer)
 
 
+def test_read_map_batches_operator_fusion_with_random_shuffle_operator(
```
Comment (c21) [inline python/ray/data/tests/test_execution_optimizer.py:610]: can we add more test cases, e.g. for multiple `map_batches` and multiple `random_shuffle`? Let's also check negative case - all other all-to-all operators except `random_shuffle` should not be fused.

File: python/ray/data/_internal/logical/rules/operator_fusion.py [hunk 5]
```diff
@@ -68,17 +119,20 @@ def _can_fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator) -> bool:
...
 
-        # We only support fusing AbstractMap -> AbstractMap operators.
-        if not isinstance(down_logical_op, AbstractMap) or not isinstance(
-            up_logical_op, AbstractMap
-        ):
+        # We currently only support fusing for the following cases:
+        # - AbstractMap -> AbstractMap
+        # - AbstractMap -> RandomShuffle
+        if not isinstance(
+            down_logical_op, (AbstractMap, RandomShuffle)
```
[the remark this answers is not in the document]
Comment (scottjlee) [reply python/ray/data/_internal/logical/rules/operator_fusion.py:129]: yes, we will do this in a future PR.

**Review by c21 — APPROVED** [review]
LGTM

**Pull request merged**